snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -133,6 +133,97 @@ def _update_yaml_with_extracted_metadata(
         raise meta_errors.MissingSnapcraftYamlKeysError(keys=missing_keys)
 
 
+def _adopt_info(
+        config_data: Dict[str, Any],
+        extracted_metadata: _metadata.ExtractedMetadata):
+    metadata_dict = extracted_metadata.to_dict()
+    for key, value in filter(
+            lambda item: item != 'desktop',
+            metadata_dict.items()):
+        if key not in config_data:
+            if key == 'icon':
+                if _icon_file_exists() or not os.path.exists(value):
+                    # Do not overwrite the icon file.
+                    continue
+            config_data[key] = value
+    if 'desktop' in metadata_dict:
+        for desktop_id in metadata_dict['desktop']:
+            app_name = _get_app_name_from_desktop_id(config_data, desktop_id)
+            if app_name and not _desktop_file_exists(app_name):
+                for desktop_file_path in (
+                        os.path.join(
+                            xdg_data_dir, 'applications',
+                            desktop_id.replace('-', '/')) for
+                        xdg_data_dir in ('usr/local/share', 'usr/share')):
```

Comment:
Another double-take here (triple, actually). If a list comprehension goes to multiple lines I get lost quickly, and here it's nested inside a `for`! Mind extracting it and making it a little easier to read?

Reply:
sure thing. This was totally unnecessary.